sympy/calculus/util.py

```diff
@@ -516,6 +516,7 @@ def _check(orig_f, period):
         g_s = decompogen(f, symbol)
         num_of_gs = len(g_s)
         if num_of_gs > 1:
+            orig_f = compogen(decompogen(orig_f, symbol), symbol) # Fix for issue 20566
```

Comment:
I don't understand what this is doing. Shouldn't this leave `orig_f` unchanged?

Is it not a bug in compogen or decompogen that this gives a different expression?

I don't think `orig_f` should be changed in any case. The name suggests that it is the original `f` (without any changes).

Reply:
> Is it not a bug in compogen or decompogen that this gives a different expression?

Not exactly, in-fact it is the simplification feature of sympy. A simple example is : -
```
from sympy import *
x = Symbol('x', real=True)
a = [exp(x), re(x), asin(x)]
print(a)
```
Gives output : 
```
[exp(x), x, asin(x)]
```
Modifying this will require changing the core simplification features of sympy. Alternatively we could also do something like:
```
 temp = Dummy('x', complex=True) 
```
On this line : 

https://github.com/sympy/sympy/blob/1e76752f545a2ed44a8af008c76c9ca7dd77e41d/sympy/calculus/util.py#L406

But that comes with it's own set of problems( We do not have support for complex periodicities in sympy and `re(x)` never gets simplified so it will simply return `None` in any `periodicity()` call by a function containing `re(x)`)

The only way I could find around this was doing `compogen(decompogen())` on the original function itself which removes all the instances of `re()`. 

> I don't think orig_f should be changed in any case. 

Then maybe this call can be made before the assignment to `orig_f` on `f` itself? 

Also `orig_f` was added as solution in #12620 for a similar problem.

Reply:
> > Is it not a bug in compogen or decompogen that this gives a different expression?
> 
> Not exactly, in-fact it is the simplification feature of sympy. A simple example is : -
> 
> ```
> from sympy import *
> x = Symbol('x', real=True)
> a = [exp(x), re(x), asin(x)]
> print(a)
> ```

I think that is a bug. We can't allow the assumptions on the symbol to affect the intermediate functions:
```python
In [88]: from sympy.solvers.decompogen import compogen

In [89]: x = Symbol('x', positive=True)

In [90]: f = abs(sin(x))

In [91]: f
Out[91]: │sin(x)│

In [92]: f2 = compogen(decompogen(f, x), x)

In [93]: f2
Out[93]: sin(x)

In [94]: f.subs(x, 4).n()
Out[94]: 0.756802495307928

In [95]: f2.subs(x, 4).n()
Out[95]: -0.756802495307928
```

Reply:
> f.subs(x, 4).n()
> 0.756802495307928

> f2.subs(x, 4).n()
> -0.756802495307928

Ah, I see, when put this way it makes sense.

Reply:
Also you can see that the periodicity of `sin(x)` and `abs(sin(x))` is not the same:
```python
In [44]: periodicity(sin(x), x)
Out[44]: 2⋅π

In [45]: periodicity(abs(sin(x)), x)
Out[45]: π
```